compiler/src/iree/compiler/Codegen/Common/DecomposeConvolutionToLowerDimOps.cpp

```diff
@@ -16,6 +19,78 @@ namespace mlir::iree_compiler {
 
 namespace {
 
+/// Computes a "decomposed" lowering config attribute for a conv OP
+///
+/// This method complements the patterns to decompose 2D convolutions into 1D
+/// convs. Specifically, it will update the lowering config attached to a Conv
+/// Op in a way that matches the "decomposition" patterns.
+///
+/// At the moment only Depthwise HWC convolutions are supported.
+static FailureOr<IREE::Codegen::LoweringConfigAttr>
+computeNewLCA(SmallVector<Operation *> computeOps, MLIRContext *context) {
+
+  // 0.1 Double-check that there's only one convolution Op among the compute Ops
+  // TODO: Make this hook work with multiple conv Ops
+  int64_t numConvOps = llvm::count_if(computeOps, [](Operation *op) {
+    return isa<linalg::ConvolutionOpInterface>(op);
+  });
+  assert(numConvOps == 1 && "Exactly 1 Linalg Conv Op is expected");
+
+  // 0.2 ATM only 2D depthwise HWC convs are supported - check that that's the
+  // one that's present.
+  // TODO: Add support for other convs
+  bool seen2DDepthwiseConvHWC = false;
+  for (auto op : computeOps) {
+    if (isa<linalg::DepthwiseConv2DNhwcHwcOp>(op))
+      seen2DDepthwiseConvHWC = true;
+  }
+  if (!seen2DDepthwiseConvHWC)
+    return failure();
+
+  // 1. Get the current lowering config attached to the Conv Op
+  FailureOr<IREE::Codegen::LoweringConfigAttr> loweringConfigAttr =
+      getLoweringConfig(computeOps);
+  if (failed(loweringConfigAttr))
+    return failure();
+
+  TilingConfig tc(loweringConfigAttr.value());
+
+  // 2. Calculate new tiling levels
+  // FIXME: This assumes that it's the H dim that's "folded". In practice, this
+  // will depend on the tiling config.
+  unsigned inputHDim = 1;
+  unsigned filterHDim = 4;
+
+  auto tilingLevels = loweringConfigAttr.value().getTilingLevels();
+  SmallVector<IREE::Codegen::LoweringConfigTilingLevelAttr> newTilingLevelsList;
+  for (auto level : tilingLevels) {
+    SmallVector<int64_t> newSizes(level.getSizes());
+    SmallVector<bool> newScalableFlags(level.getScalableFlags());
+
+    newSizes.erase(newSizes.begin() + inputHDim);
+    newSizes.erase(newSizes.begin() + filterHDim);
+    if (newScalableFlags.size() > 0) {
+      newScalableFlags.erase(newScalableFlags.begin() + inputHDim);
+      newScalableFlags.erase(newScalableFlags.begin() + filterHDim);
+    }
+
+    SmallVector<int64_t> interchange = {};
+    auto newLevel = IREE::Codegen::LoweringConfigTilingLevelAttr::get(
+        context, newSizes, ArrayRef<int64_t>{}, newScalableFlags);
```

Comment:
Likely not relevant here, but note that this is dropping the interchange.

Reply:
I've not seen `interchange` being used much in KernelDispatch.cpp, let alone for Convs specifically. @hanhanW Do you know what transformations might be setting it up (or using)?

Reply:
Maybe just fail the pattern if interchange is present rather than silently drop it?

Reply:
+1 on failing the pattern if it is present.

The interchange path was experimental. We added the features because it is natively supported in LinAlg tiling. However, we did not find it helpful in our workload. So it is not used in KernelDispatch.cpp.

Today, we have few tests in the distribution pass, and there is an e2e tests in https://github.com/iree-org/iree/blob/62950743a899dad13aeb502c146b352a869c8fad/tests/e2e/regression/lowering_config.mlir#L7-L9

Reply:
nit: add a function argument comment. (And we can potentially drop `ArrayRef<>`)

```suggestion
        context, newSizes, /*interchange=*/{}, newScalableFlags);
```